cql2pgjson/src/main/java/org/folio/cql2pgjson/CQL2PgJSON.java

```diff
@@ -765,27 +765,58 @@ private String pgId(CQLTermNode node) throws QueryValidationException {
 
+    final Index schemaIndex = DbSchemaUtils.getIndex(index, this.dbTable.getFullTextIndex());
+    IndexTextAndJsonValues vals = new IndexTextAndJsonValues();
     StringBuilder res = new StringBuilder();
+    if (schemaIndex != null) {
+      final String modifiersSubfield = schemaIndex.getModifiersSubfield();
+      if (modifiersSubfield != null) {
+        vals.setIndexText(index2sqlText("t.c", modifiersSubfield));
 
-    IndexTextAndJsonValues vals = new IndexTextAndJsonValues();
-    vals.setIndexText(index2sqlText("t.c", relationModifier.getSubfield()));
-
-    final String table = this.jsonField.split("\\.")[0];
-    final String field = this.jsonField.split("\\.")[1];
-
-    res.append("id in (select t.id from (select id as id, jsonb_array_elements(");
-    res.append(field + "->'" + index + "') as c from " + table + ") as t where t.c");
-    res.append(" @> '{\"");
-    // pass the name from the schema rather than CQL because case is lost
-    res.append(Cql2SqlUtil.cql2string(relationModifier.getModifierName()));
-    res.append("\": \"");
-    res.append(Cql2SqlUtil.cql2string(modifierValue));
-    res.append("\"}' and ");
-    res.append(indexNode(index, node, vals, dbIndex, modifiers));
-    res.append(")");
+        final String table = this.jsonField.split("\\.")[0];
+        final String field = this.jsonField.split("\\.")[1];
+
+        res.append("id in (select t.id from (select id as id, jsonb_array_elements(");
+        res.append(field + "->'" + index + "') as c from " + table + ") as t where t.c");
+        res.append(" @> '{");
+      }
+    }
+    int no = 0;
+    for (Modifier relationModifier : relationModifiers) {
+      final String modifierName = relationModifier.getType().substring(1);
+      final String modifierValue = relationModifier.getValue();
+      if (modifierValue == null) {
+        throw new QueryValidationException("CQL: Missing value for relation modifier " + relationModifier.getType());
+      }
+      if (!"=".equals(relationModifier.getComparison())) {
+        throw new QueryValidationException("CQL: Unsupported comparison for relation modifier " + relationModifier.getType());
+      }
+      boolean found = false;
+      if (schemaIndex != null) {
+        List<String> schemaModifiers = schemaIndex.getModifiers();
+        if (schemaModifiers != null) {
+          for (String schemaModifier : schemaModifiers) {
+            if (schemaModifier.equalsIgnoreCase(modifierName)) {
+              if (no > 0) {
+                res.append(',');
+              }
+              res.append('\"').append(Cql2SqlUtil.cql2string(schemaModifier))
+                .append("\": \"").append(Cql2SqlUtil.cql2string(modifierValue))
+                .append("\"");
+              no++;
+              found = true;
+            }
+          }
+        }
+      }
+      if (!found) {
+        throw new QueryValidationException("CQL: Unsupported relation modifier " + relationModifier.getType());
+      }
+    }
```

Comment:
The code for the right hand side of `@>` should be moved into a subfunction.

Reply:
It could and SonarQube will warn if the function is too complex.